k8s/src/main/scala/io/buoyant/k8s/resources.scala

```diff
@@ -0,0 +1,308 @@
+package io.buoyant.k8s
+
+import com.twitter.concurrent.AsyncStream
+import com.twitter.finagle.http.Response
+import com.twitter.finagle.{Filter, http}
+import com.twitter.finagle.param.HighResTimer
+import com.twitter.finagle.service.{Backoff, RetryBudget, RetryFilter, RetryPolicy}
+import com.twitter.finagle.stats.{DefaultStatsReceiver, StatsReceiver}
+import com.twitter.io.{Buf, Reader}
+import com.twitter.util.{Closable, _}
+import com.twitter.util.TimeConversions._
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Basic functionality shared by all k8s Resource implementations: they require an HTTP client, and can be closed.
+ */
+trait Resource extends Closable {
+  def client: Client
+  def close(deadline: Time) = client.close(deadline)
+}
+
+/**
+ * A Version is a top-level API grouping within the K8s API; examples include "/api/v1" and "/apis/extensions/v1beta1".
+ *
+ * Implementors are likely to want to extend this to provide named convenience methods on top of the `list` method and
+ * to type-specialize on the resource types supported by that version.
+ *
+ * @tparam O The parent type for all [Object]s served by this version.
+ */
+private[k8s] trait Version[O <: Object] extends Resource {
+  /**
+   * The first portion of the API path. Currently-known groups are "api" (for the core k8s v1 API) and "apis"
+   * (all others, including extensions and third-party resources).
+   */
+  def group: String
+
+  /**
+   * The second portion of the API path. examples include "v1", "extensions/v1beta1", "buoyant.io/v1".
+   */
+  def version: String
+
+  def client: Client
+
+  def namespace(ns: String) = new NsVersion[O](client, group, version, ns)
+  def path = s"/$group/$version"
+  def list[T <: O: Manifest, W <: Watch[T]: Manifest, L <: List[T]: Manifest](
```

Comment:
perhaps `listResource` would be clearer? it was slightly odd to see `dtabs = list[...]`